keps/262-ConfigurableFailurePolicy/README.md

```diff
@@ -0,0 +1,517 @@
+              values: [143] # SIGTERM = exit code 143
+        template:
+          spec:
+            restartPolicy: Never
+            containers:
+            - name: main
+              image: bash:latest
+              image: docker.io/library/bash:5
+              command: ["bash"]
+              args:
+              - -c
+              - echo "Hello world! I'm going to exit with exit code 143 (SIGTERM) to simulate host maintenance." && sleep 20 && exit 143
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Proposed Failure Policy API
+
+```go
+
+// FailurePolicyAction defines the action the JobSet controller will take for
+// a given FailurePolicyRule.
+type FailurePolicyAction string
+
+const (
+  // Fail JobSet immediately, regardless of maxRestarts.
+  FailJobSet FailurePolicyAction = "FailJobSet"
+
+  // Don't count the failure against maxRestarts.
+  Ignore FailurePolicyAction = "Ignore"
+)
+
+// FailurePolicyRule defines a FailurePolicyAction to be executed if a child job
+// fails due to a reason listed in OnJobFailureReasons.
+type FailurePolicyRule struct {
+  // The action to take if the rule is matched.
+  // +kubebuilder:validation:Enum:=FailJobSet;Ignore;FailJob;RestartJob
+  Action FailurePolicyAction `json:"action"`
+  // The requirement on the job failure reasons. The requirement
+  // is satisfied if at least one reason matches the list.
+  // The rules are evaluated in order, and the first matching
+  // rule is executed.
+  // An empty list applies the rule to any job failure reason.
+  // +kubebuilder:validation:UniqueItems:true
+  OnJobFailureReasons []string `json:"onJobFailureReasons"`
+  // TargetReplicatedJobs are the names of the replicated jobs the operator applies to.
+  // An empty list will apply to all replicatedJobs.
+  // +optional
+  // +listType=atomic
+  // +kubebuilder:validation:UniqueItems
+  TargetReplicatedJobs []string `json:"targetReplicatedJobs,omitempty"`
+}
+
+type FailurePolicy struct {
+  // MaxRestarts defines the limit on the number of JobSet restarts.
+  // A restart is achieved by recreating all active child jobs.
+  MaxRestarts int32 `json:"maxRestarts,omitempty"`
+  // List of failure policy rules for this JobSet.
+  // For a given Job failure, the rules will be evaluated in order,
+  // and only the first matching rule will be executed.
```

Comment:
add that if no rule matches, then the default is to RestartJobSet

Reply:
Updated